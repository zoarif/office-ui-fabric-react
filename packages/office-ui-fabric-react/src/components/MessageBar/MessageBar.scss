@import '../../common/common';

// Copyright (c) Microsoft. All rights reserved. Licensed under the MIT license. See LICENSE in the project root for license information.

//
// Office UI Fabric
// --------------------------------------------------
// MessageBar Styles

// Mixins
@mixin multiLineContent {
  flex-direction: column;
  .content {
    flex-direction: row;
  }
  .innerText {
    overflow: visible;
    white-space: pre-wrap;
  }
}

@mixin buttonMargins($sidesAndBottom, $top: $sidesAndBottom) {
  @include margin($top, $sidesAndBottom, $sidesAndBottom, 0);
  :nth-child(n+2) {
    @include margin-left($sidesAndBottom);
  }
}

.root {
  @include ms-bgColor-info;
  min-height: 32px;
  color: $ms-color-infoText;
  width: 100%;
  box-sizing: border-box;
  display: flex;
  position: relative;
  word-break: break-word;

  :global(.ms-Link) {
    font-size: $ms-font-size-s;
    color: $ms-color-themeDark;
  }
}

.icon,
.text {
  display: table-cell;
  vertical-align: top;
}

.icon {
  font-size: $ms-icon-size-m;
  min-width: 16px;
  min-height: 16px;
  display: flex;
  color: $ms-color-info;
  flex-shrink: 0;
  // Margin-right is accounted for in the .innerText margin-left
  @include margin(16px, 0, 16px, 16px);
  @media screen and (max-width: $ms-screen-max-sm) {
    @include margin(8px, 0, 8px, 8px);
  }
}

.text {
  @include ms-font-s;
  min-width: 0;
  display: flex;
  flex-grow: 1;
  @include margin(16px, 8px, 16px, 8px);
  @media screen and (max-width: $ms-screen-max-sm) {
    @include margin(8px, 0, 8px, 8px);
    &.noDismissButton {
      @include margin-right(8px);
    }
  }
  &.noDismissButton {
    @include margin-right(16px);
  }
}

.innerText {
  line-height: 16px;

  // Add padding between text and hyperlink
  span a {
    @include padding-left(4px);
  }
}

//== Modifier: Warning message
//
.root.rootIsWarning {
  @include ms-bgColor-warning;
  color: $ms-color-warningText;
}


//== Modifier: Severe Warning message
//
.root.rootIsSevereWarning {
  @include ms-bgColor-severeWarning;
  color: $ms-color-severeWarningText;

  .icon {
    @include ms-fontColor-severeWarning;
  }
}


//== Modifier: Error message
//
.root.rootIsError {
  @include ms-bgColor-error;
  color: $ms-color-errorText;

  .icon {
    @include ms-fontColor-error;
  }
}


//== Modifier: Blocked message
//
.root.rootIsBlocked {
  @include ms-bgColor-error;
  color: $ms-color-errorText;

  .icon {
    @include ms-fontColor-error;
  }
}


//== Modifier: Success message
//
.root.rootIsSuccess {
  @include ms-bgColor-success;
  color: $ms-color-successText;

  .icon {
    @include ms-fontColor-green;
  }
}

// Shared between Single- and Multi- Line
.content {
  display: flex;
  line-height: normal;
  width: 100%;
  box-sizing: border-box;
  &:before {
    pointer-events: none;
    position: absolute;
    right: 0;
    bottom: 0;
    left: 0;
    top: 0;
    margin: 0;
    @include high-contrast {
        border: 1px solid WindowText;
        content: ' ';
    }
  }
}

.actions {
  @include buttonMargins(12px, 0);
}

.actionsSingleLine {
  @include buttonMargins(8px);
}

.actions, .actionsSingleLine {
  display: flex;
  flex: 0 0 auto;
  flex-direction: row-reverse;
  align-items: center;
  @media screen and (max-width: $ms-screen-max-sm) {
    @include buttonMargins(8px, 0);
    :global(.ms-Button) {
      margin-top: 8px;
    }
  }
}

.dismissal,
.expand {
  @include focus-border();
  flex-shrink: 0;
  @include margin(8px, 8px, 8px, 0); // Margin-left is accounted for in the .text margin-right

  @media screen and (max-width: $ms-screen-max-sm) {
    @include margin(0, 0, 0, 8px);
  }

  @include high-contrast {
    -ms-high-contrast-adjust: none;
  }
}

.dismissSingleLine,
.expandSingleLine {
  display: flex;
}

.dismissal,
.dismissSingleLine,
.expandSingleLine {
  :global(.ms-Button-icon) {
    font-size: 12px;
    height: 12px;
    line-height: 12px;
    color: $ms-color-neutralSecondary;
  }
}

// Multi line specific attributes
.multiLine {
  @include multiLineContent();
  .text.multiLineWithActions {
    margin-bottom: 8px; // This creates an 8px space between text and actions
    @media screen and (max-width: $ms-screen-max-sm) {
      margin-bottom: 0;
    }
  }
}

// Single line specific attributes
.singleLine {
  .innerText {
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }
  .expandSingleLine {
    @include multiLineContent();
  }
  @media screen and (max-width: $ms-screen-max-sm) {
    // Singleline acts like multiLine in mobile, unless text is specified to be truncated
    &:not(.expandingSingleLine) {
      @include multiLineContent();
      .text.multiLineWithActions {
        @media screen and (max-width: $ms-screen-max-sm) {
          margin-bottom: 0;
        }
      }
    }
  }
}