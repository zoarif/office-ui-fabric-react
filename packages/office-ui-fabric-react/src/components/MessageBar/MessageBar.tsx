import * as React from 'react';
import {
  BaseComponent,
  DelayedRender,
  css,
  getId
} from '../../Utilities';
import { IconButton } from '../../Button';
import { Icon } from '../../Icon';
import { IMessageBarProps, MessageBarType } from './MessageBar.types';
import * as stylesImport from './MessageBar.scss';
const styles: any = stylesImport;

export interface IMessageBarState {
  labelId?: string;
  showContent?: boolean;
  expandSingleLine?: boolean;
}

export class MessageBar extends BaseComponent<IMessageBarProps, IMessageBarState> {

  public static defaultProps: IMessageBarProps = {
    messageBarType: MessageBarType.info,
    onDismiss: undefined,
    isMultiline: true,
  };

  private ICON_MAP = {
    [MessageBarType.info]: 'Info',
    [MessageBarType.warning]: 'Info',
    [MessageBarType.error]: 'ErrorBadge',
    [MessageBarType.blocked]: 'Blocked2',
    [MessageBarType.remove]: 'Blocked', // TODO remove deprecated value at >= 1.0.0
    [MessageBarType.severeWarning]: 'Warning',
    [MessageBarType.success]: 'Completed'
  };

  constructor(props: IMessageBarProps) {
    super(props);

    this.state = {
      labelId: getId('MessageBar'),
      showContent: false,
      expandSingleLine: false,
    };
  }

  public render(): JSX.Element {
    const { isMultiline } = this.props;

    return isMultiline ? this._renderMultiLine() : this._renderSingleLine();
  }

  private _getActionsDiv(): JSX.Element | null {
    if (this.props.actions) {
      return (
        <div
          className={ this.props.isMultiline ?
            ('ms-MessageBar-actions ' + styles.actions) :
            ('ms-MessageBar-actionsSingleLine ' + styles.actionsSingleLine) }
        >
          { this.props.actions }
        </div>
      );
    }
    return null;
  }

  private _getClassName(): string {
    return css(this.props.className, 'ms-MessageBar', styles.root, {
      ['ms-MessageBar ' + styles.root]: this.props.messageBarType === MessageBarType.info,
      ['ms-MessageBar--error ' + styles.rootIsError]: this.props.messageBarType === MessageBarType.error,
      ['ms-MessageBar--blocked ' + styles.rootIsBlocked]: (this.props.messageBarType === MessageBarType.blocked) ||
        (this.props.messageBarType === MessageBarType.remove), // TODO remove deprecated value at >= 1.0.0
      ['ms-MessageBar--severeWarning ' + styles.rootIsSevereWarning]: this.props.messageBarType === MessageBarType.severeWarning,
      ['ms-MessageBar--success ' + styles.rootIsSuccess]: this.props.messageBarType === MessageBarType.success,
      ['ms-MessageBar--warning ' + styles.rootIsWarning]: this.props.messageBarType === MessageBarType.warning
    });
  }

  private _getDismissDiv(): JSX.Element | null {
    if (this.props.onDismiss) {
      return (
        <IconButton
          disabled={ false }
          className={ css('ms-MessageBar-dismissal', styles.dismissal) }
          onClick={ this.props.onDismiss }
          iconProps={ { iconName: 'Clear' } }
          ariaLabel={ this.props.dismissButtonAriaLabel }
        />
      );
    }
    return null;
  }

  private _getDismissSingleLine(): JSX.Element | null {
    if (this.props.onDismiss) {
      return (
        <div className={ css('ms-MessageBar-dismissSingleLine', styles.dismissSingleLine) }>
          { this._getDismissDiv() }
        </div>
      );
    }
    return null;
  }

  private _getExpandSingleLine(): JSX.Element | null {
    if (!this.props.actions && this.props.truncated) {
      return (
        <div className={ css('ms-MessageBar-expandSingleLine', styles.expandSingleLine) }>
          <IconButton
            disabled={ false }
            className={ css('ms-MessageBar-expand', styles.expand) }
            onClick={ this._onClick }
            iconProps={ { iconName: this.state.expandSingleLine ? 'DoubleChevronUp' : 'DoubleChevronDown' } }
            ariaLabel={ this.props.overflowButtonAriaLabel }
          />
        </div>
      );
    }
    return null;
  }

  private _getIconSpan(): JSX.Element {
    return (
      <div className={ css('ms-MessageBar-icon', styles.icon) }>
        <Icon iconName={ this.ICON_MAP[this.props.messageBarType!] } />
      </div>
    );
  }

  private _renderMultiLine(): React.ReactElement<React.HTMLAttributes<HTMLAreaElement>> {
    return (
      <div
        className={
          css(this._getClassName(),
            'ms-MessageBar-multiline',
            styles.multiLine,
            this.props.onDismiss && styles.dismissalMultiLine
          )
        }
        role='status'
        aria-live={ this._getAnnouncementPriority() }
      >
        <div className={ css(styles.content, 'ms-MessageBar-content') }>
          { this._getIconSpan() }
          { this._renderInnerText() }
          { this._getDismissDiv() }
        </div>
        { this._getActionsDiv() }
      </div >
    );
  }

  private _renderSingleLine(): React.ReactElement<React.HTMLAttributes<HTMLAreaElement>> {
    return (
      <div
        className={
          css(this._getClassName(),
            'ms-MessageBar-singleline',
            styles.singleLine,
            this.props.onDismiss && 'ms-MessageBar-dismissalSingleLine',
            this.props.truncated && 'ms-MessageBar-expandingSingleLine',
            this.props.truncated && styles.expandingSingleLine
          )
        }
      >
        <div className={ css(styles.content, 'ms-MessageBar-content') }>
          { this._getIconSpan() }
          { this._renderInnerText() }
          { this._getExpandSingleLine() }
          { this._getActionsDiv() }
          { this._getDismissSingleLine() }
        </div>
      </div >
    );
  }

  private _renderInnerText(): JSX.Element {
    return (
      <div
        className={ css('ms-MessageBar-text', styles.text, this.props.actions && styles.multiLineWithActions, !this.props.onDismiss && styles.noDismissButton, this.state.expandSingleLine && styles.expandSingleLine) }
        id={ this.state.labelId }
      >
        <span
          className={ css('ms-MessageBar-innerText ' + styles.innerText) }
          role='status'
          aria-live={ this._getAnnouncementPriority() }
        >
          <DelayedRender>
            <span>{ this.props.children }</span>
          </DelayedRender>
        </span>
      </div >
    );
  }

  private _getAnnouncementPriority(): 'assertive' | 'polite' {
    switch (this.props.messageBarType) {
      case MessageBarType.blocked:
      case MessageBarType.error:
      case MessageBarType.severeWarning:
        return 'assertive';
    }
    return 'polite';
  }

  private _onClick = (ev: React.MouseEvent<HTMLAnchorElement | HTMLButtonElement>) => {
    this.setState({ expandSingleLine: !this.state.expandSingleLine });
  }
}