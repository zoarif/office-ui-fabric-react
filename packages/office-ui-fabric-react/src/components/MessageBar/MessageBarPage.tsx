import * as React from 'react';
import {
  ExampleCard,
  IComponentDemoPageProps,
  ComponentPage,
  PageMarkdown,
  PropertiesTableSet
} from '@uifabric/example-app-base';
import { MessageBarBasicExample } from './examples/MessageBar.Basic.Example';
import { ComponentStatus } from '../../demo/ComponentStatus/ComponentStatus';
import { MessageBarStatus } from './MessageBar.checklist';

const MessageBarBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/MessageBar/examples/MessageBar.Basic.Example.tsx') as string;

export class MessageBarPage extends React.Component<IComponentDemoPageProps, {}> {
  public render(): JSX.Element {
    return (
      <ComponentPage
        title='MessageBar'
        componentName='MessageBarExample'
        componentUrl='https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/MessageBar'
        exampleCards={
          <ExampleCard
            title='Various MessageBar types'
            code={ MessageBarBasicExampleCode }
          >
            <MessageBarBasicExample />
          </ExampleCard>
        }
        propertiesTables={
          <PropertiesTableSet
            sources={ [
              require<string>('!raw-loader!office-ui-fabric-react/src/components/MessageBar/MessageBar.types.ts')
            ] }
          />
        }
        overview={
          <PageMarkdown>
            { require<string>('!raw-loader!office-ui-fabric-react/src/components/MessageBar/docs/MessageBarOverview.md') }
          </PageMarkdown>
        }
        bestPractices={
          <div />
        }
        dos={
          <PageMarkdown>
            { require<string>('!raw-loader!office-ui-fabric-react/src/components/MessageBar/docs/MessageBarDos.md') }
          </PageMarkdown>
        }
        donts={
          <PageMarkdown>
            { require<string>('!raw-loader!office-ui-fabric-react/src/components/MessageBar/docs/MessageBarDonts.md') }
          </PageMarkdown>
        }
        isHeaderVisible={ this.props.isHeaderVisible }
        componentStatus={
          <ComponentStatus
            { ...MessageBarStatus }
          />
        }
      />
    );
  }

}
