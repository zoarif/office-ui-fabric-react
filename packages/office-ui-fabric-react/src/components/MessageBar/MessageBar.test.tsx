import * as React from 'react';
import * as ReactDOM from 'react-dom';
import * as ReactTestUtils from 'react-dom/test-utils';
import * as renderer from 'react-test-renderer';

import { MessageBar } from './MessageBar';

describe('MessageBar', () => {
  const noop = () => {
    /* no-op */
  };

  function renderIntoDocument(element: React.ReactElement<any>): HTMLElement {
    const component = ReactTestUtils.renderIntoDocument(element);
    const renderedDOM = ReactDOM.findDOMNode(component as React.ReactInstance);
    return renderedDOM as HTMLElement;
  }

  it('renders MessageBar correctly', () => {
    const component = renderer.create(<MessageBar>Message</MessageBar>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('dismiss', () => {
    describe('single-line', () => {
      it('is present when onDismiss exists', () => {
        const renderedDOM: HTMLElement = renderIntoDocument(<MessageBar onDismiss={ noop } isMultiline={ false } />);
        const dismissElement = renderedDOM.querySelector('.ms-MessageBar-dismissal');
        expect(dismissElement).not.toBeNull();
      });

      it('is not present when onDismiss is missing', () => {
        const renderedDOM: HTMLElement = renderIntoDocument(<MessageBar isMultiline={ false } />);
        const dismissElement = renderedDOM.querySelector('.ms-MessageBar-dismissal');
        expect(dismissElement).toBeNull();
      });
    });

    describe('multi-line', () => {
      it('is present when onDismiss exists', () => {
        const renderedDOM: HTMLElement = renderIntoDocument(<MessageBar onDismiss={ noop } isMultiline={ true } />);
        const dismissElement = renderedDOM.querySelector('.ms-MessageBar-dismissal');
        expect(dismissElement).not.toBeNull();
      });

      it('is not present when onDismiss is missing', () => {
        const renderedDOM: HTMLElement = renderIntoDocument(<MessageBar isMultiline={ true } />);
        const dismissElement = renderedDOM.querySelector('.ms-MessageBar-dismissal');
        expect(dismissElement).toBeNull();
      });
    });
  });

  describe('truncated', () => {
    it('is present when onDismiss exists', () => {
      const renderedDOM: HTMLElement = renderIntoDocument(<MessageBar truncated={ true } isMultiline={ false } />);
      const expandElement = renderedDOM.querySelector('.ms-MessageBar-expand');
      expect(expandElement).not.toBeNull();
    });

    it('is not present when truncated is missing', () => {
      const renderedDOM: HTMLElement = renderIntoDocument(<MessageBar isMultiline={ false } />);
      const expandElement = renderedDOM.querySelector('.ms-MessageBar-expand');
      expect(expandElement).toBeNull();
    });
  });
});
