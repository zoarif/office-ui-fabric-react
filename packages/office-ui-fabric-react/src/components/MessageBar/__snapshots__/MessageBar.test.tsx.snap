// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageBar renders MessageBar correctly 1`] = `
<div
  aria-live="polite"
  className="ms-MessageBar ms-MessageBar undefined ms-MessageBar-multiline"
  role="status"
>
  <div
    className="ms-MessageBar-content"
  >
    <div
      className="ms-MessageBar-icon"
    >
      <i
        aria-hidden={true}
        className=

            {
              display: inline-block;
            }
        data-icon-name="Info"
        role="presentation"
      />
    </div>
    <div
      className="ms-MessageBar-text"
      id="MessageBar0"
    >
      <span
        aria-live="polite"
        className="ms-MessageBar-innerText undefined"
        role="status"
      />
    </div>
  </div>
</div>
`;
